torch/csrc/jit/codegen/cuda/expr_simplifier.cpp

```diff
@@ -696,8 +718,219 @@ bool isFlattenedMul(Val* x) {
   return toFlattenedMul(x->definition()) != nullptr;
 }
 
+std::pair<int64_t, std::list<Val*>> getConstAndSymbolicFactors(Val* x) {
+  std::vector<Val*> factors;
+  if (auto fop = toFlattenedMul(x->definition())) {
+    factors = fop->inputs();
+  } else {
+    factors.emplace_back(x);
+  }
+  int64_t const_factor = 1;
+  std::list<Val*> symbolic_factors;
+  for (auto f : factors) {
+    f = foldConstants(f);
+    if (f->getInt().has_value()) {
+      const_factor *= *f->getInt();
+    } else {
+      symbolic_factors.emplace_back(f);
+    }
+  }
+  return {const_factor, symbolic_factors};
+}
+
+Val* productOfFactors(
+    int64_t const_factor,
+    std::vector<Val*> symbolic_factors,
+    DataType dtype) {
+  if (const_factor != 1) {
+    symbolic_factors.emplace_back(IrBuilder::newConstant(const_factor, dtype));
+  }
+  if (symbolic_factors.size() == 1) {
+    return symbolic_factors.at(0);
+  }
+  if (symbolic_factors.empty()) {
+    return IrBuilder::newConstant(1, dtype);
+  }
+  auto output = IrBuilder::newScalar(dtype);
+  IrBuilder::create<FOp>(
+      BinaryOpType::Mul, output, std::move(symbolic_factors));
+  return output;
+}
+
 } // namespace
 
+namespace sym_algebra {
+
+// Common utilities for symbolic algebra.
+
+// Rewrite x in the form x = x1 * x2 * x3 * ...
+Val* factorize(Val* x);
+
+// Given that x = x1 * x2 * x3 * x4 * ..., y = x2 * x4 * ..., where x is a
+// multiple of y, evaluate x/y as x/y = x1 * x3 * ... Both x and y should have
+// already been factorized in the form of products, otherwise, this function
+// might not be able to get the desired result. Returns nullptr if not
+// divisible. Note that this function does symbolic term cancellation, it does
+// not require y to be non-zero.
+Val* divideFactorized(Val* x, Val* y) {
+  auto x_factors = getConstAndSymbolicFactors(x);
+  auto y_factors = getConstAndSymbolicFactors(y);
+
+  int64_t quoient_const_factor;
+  std::vector<Val*> quoient_symbolic_factors;
+
+  if (x_factors.first % y_factors.first != 0) {
+    // not divisible
+    return nullptr;
+  } else {
+    quoient_const_factor = x_factors.first / y_factors.first;
+  }
+
+  for (auto yf : y_factors.second) {
```

Comment:
Seems similar to:
https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/scheduler/vectorize_helper.cpp#L110-L126

Reply:
Yeah, I should do some refactoring in a separate PR.